File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/numberpath/domain/NumberPathGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.numberpath.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.List;

import fj.F2;

/**
 * Created by zwvista on 2016/09/29.
 */

public class NumberPathGame extends CellsGame<NumberPathGame, NumberPathGameMove, NumberPathGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };

    public int[] objArray;
    public int get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public int get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, int obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, int obj) {
        set(p.row, p.col, obj);
    }

    public NumberPathGame(List<String> layout, GameInterface<NumberPathGame, NumberPathGameMove, NumberPathGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length() / 2);
        objArray = new int[rows() * cols()];
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                String s = str.substring(c * 2, c * 2 + 2).trim();
                set(p, Integer.valueOf(s));
            }
        }
        NumberPathGameState state = new NumberPathGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(NumberPathGameMove move, F2<NumberPathGameState, NumberPathGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        NumberPathGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean setObject(NumberPathGameMove move) {
        return changeObject(move, (state, move2) -> state.setObject(move2));
    }

    public Boolean[] getObject(Position p) {
        return state().get(p);
    }

    public Boolean[] getObject(int row, int col) {
        return state().get(row, col);
    }

    public HintState pos2State(Position p) {
        return state().pos2state.get(p);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/numberpath/domain/NumberPathGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.numberpath.domainimport

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState
import fj.F2

class NumberPathGame(layout: List<String>, gi: GameInterface<NumberPathGame, NumberPathGameMove, NumberPathGameState>, gdi: GameDocumentInterface) : CellsGame<NumberPathGame, NumberPathGameMove, NumberPathGameState>(gi, gdi) {
    var objArray: IntArray
    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]

    operator fun get(p: Position) = get(p.row, p.col)

    operator fun set(row: Int, col: Int, obj: Int) {
        objArray[row * cols() + col] = obj
    }

    operator fun set(p: Position, obj: Int) {
        set(p.row, p.col, obj)
    }

    private fun changeObject(move: NumberPathGameMove, f: (NumberPathGameState, NumberPathGameMove) -> Boolean): Boolean {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state: NumberPathGameState = cloner.deepClone(state())
        val changed = f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states.get(stateIndex - 1), state)
        }
        return changed
    }

    fun setObject(move: NumberPathGameMove) = changeObject(move, F2<NumberPathGameState, NumberPathGameMove, Boolean> { state: NumberPathGameState, move2: NumberPathGameMove -> state.setObject(move2) })

    fun getObject(p: Position?) = state().get(p)

    fun getObject(row: Int, col: Int) = state().get(row, col)

    fun pos2State(p: Position?) = state().pos2state.get(p)

    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1))
    }

    init {
        size = Position(layout.size, layout[0].length / 2)
        objArray = IntArray(rows() * cols())
        for (r in 0 until rows()) {
            val str = layout[r]
            for (c in 0 until cols()) {
                val p = Position(r, c)
                val s = str.substring(c * 2, c * 2 + 2).trim { it <= ' ' }
                set(p, Integer.valueOf(s))
            }
        }
        val state = NumberPathGameState(this)
        states.add(state)
        levelInitilized(state)
    }
}